doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1422 @@
+
+#### 2.2.3.1 Full Cone NAT
+When the first connection from a particular internal IP + L4 port is dynamically SNAT mapped to an external IP + L4 port, the SNAT and the corresponding DNAT entries are added for that mapping in the hardware.
+Subsequently any new connections from the same internal IP + L4 port to new destination end points use the same mapping. As well, traffic sent by any external host to this external IP + L4 port is NAT'ted back to the same internal IP + L4 port. This behavior model is referred to as Full Cone NAT.
+
+### 2.2.4 NAT zones
+NAT zones refer to different network domains between which the NAT translation happens when the packet crosses between them. NAT zones are also referred to as NAT realms.
+NAT zones are created by configuring zone-id per L3 interface. The L3 interface referred to for NAT purposes can be an Ethernet, VLAN or PortChannel interface that are configured with IP address(es).
+
+In this document, the interface that is towards the private networks (private realm) on the NAT router is referred to as an inside interface.
+And the interface that is towards the public network (public realm) on the NAT router is referred to as outside interface.
+By default, L3 interface is in NAT zone 0 which we refer to as an inside interface.
+
+NAT/NAPT is performed when packets matching the configured NAT ACLs cross between different zones.
+The source zone of a packet is determined by the zone of the interface on which the packet came on. And the destination zone of the packet is determined by the zone of the L3 next-hop interface from the L3 route lookup of the destination.
+
+Currently only 2 zones are supported, which correspond to the inside interface and the outside interface.
+
+Any inbound traffic ingressing on the outside interface that is L3 forwarded on to the inside interface, is configured by user via Static NAT/NAPT entries to be DNAT translated.
+Any outbound traffic ingressing on the inside interface is configured to be dynamically SNAT translated.
+
+The public IP address of the interface deemed to be outside interface, is referred to in the NAT pool configuration.
+
+The allowed range of configurable zone values is 0 to 3.
+
+### 2.2.5 Twice NAT and NAPT
+Twice NAT or Double NAT is a NAT variation where both the Source IP and the Destination IP addresses are modified as a packet crosses the address zones. It is typically used in  the communication between networks with overlapping private addresses.
+
+The configuration for Twice NAT/NAPT is achieved in 2 ways:
+- Putting two Static NAT/NAPT entries in the same Group ('twice_nat_id' value).
+- Putting a Dynamic NAT/NAPT Binding and a Static type NAT/NAPT entry in the same Group ('twice_nat_id' value).
+
+When a host matching a dynamic NAT pool binding sends traffic to host with a matching DNAT Static NAT/NAPT entry in the same 'twice_nat_id' group, a bi-directional Twice NAT/NAPT entry is created for the traffic flow.
```

Comment:
does linux support twice nat scenarios, what linux configuration do we need.

Reply:
No additional configuration is needed to support the twice NAT.

In the kernel, when the packet is traversing from the ingress to the egress:
1. It creates a conntrack entry if it doesn't exist
2. It first hits the PREROUTING rules in iptables, does the DNAT and updates the conntrack entry with the translated DIP+DPORT if matched
3. L3 routes the packet
4. Next it hits the POSTROUTING rules in iptables, does the SNAT and updates the conntrack entry with the translated SIP+SPORT if matched

We get netlink notifications in the NAT application whenever the conntrack entry is modified. After step 4, we see that the conntrack 5-tuple is modified with both the SNAT and DNAT actions. It implies the Twice NAT happened.